extensions/rv32-adapters/src/eq_mod.rs

```diff
@@ -227,211 +222,212 @@ impl<
+    ) {
         let Instruction { a, d, .. } = *instruction;
-        let (rd_id, _) = memory.write(d, a, &tmp_convert_to_u8s(output.writes[0]));
-
-        debug_assert!(
-            memory.timestamp() - from_state.timestamp
-                == (NUM_READS * (BLOCKS_PER_READ + 1) + 1) as u32,
-            "timestamp delta is {}, expected {}",
-            memory.timestamp() - from_state.timestamp,
-            NUM_READS * (BLOCKS_PER_READ + 1) + 1
+        let cols: &mut Rv32IsEqualModAdapterCols<F, NUM_READS, BLOCKS_PER_READ, BLOCK_SIZE> =
+            adapter_row.borrow_mut();
+        cols.rd_ptr = a;
+        tracing_write(
+            memory,
+            d.as_canonical_u32(),
+            a.as_canonical_u32(),
+            data,
+            &mut cols.writes_aux,
         );
-
-        Ok((
-            ExecutionState {
-                pc: from_state.pc + DEFAULT_PC_STEP,
-                timestamp: memory.timestamp(),
-            },
-            Self::WriteRecord { from_state, rd_id },
-        ))
     }
 
-    fn generate_trace_row(
+    fn fill_trace_row(
         &self,
-        row_slice: &mut [F],
-        read_record: Self::ReadRecord,
-        write_record: Self::WriteRecord,
-        memory: &OfflineMemory<F>,
+        mem_helper: &MemoryAuxColsFactory<F>,
+        _ctx: (),
+        adapter_row: &mut [F],
     ) {
-        let aux_cols_factory = memory.aux_cols_factory();
-        let row_slice: &mut Rv32IsEqualModAdapterCols<F, NUM_READS, BLOCKS_PER_READ, BLOCK_SIZE> =
-            row_slice.borrow_mut();
-        row_slice.from_state = write_record.from_state.map(F::from_canonical_u32);
-
-        let rs = read_record.rs.map(|r| memory.record_by_id(r));
-        for (i, r) in rs.iter().enumerate() {
-            row_slice.rs_ptr[i] = r.pointer;
-            row_slice.rs_val[i].copy_from_slice(r.data_slice());
-            aux_cols_factory.generate_read_aux(r, &mut row_slice.rs_read_aux[i]);
-            for (j, x) in read_record.reads[i].iter().enumerate() {
-                let read = memory.record_by_id(*x);
-                aux_cols_factory.generate_read_aux(read, &mut row_slice.heap_read_aux[i][j]);
-            }
+        let cols: &mut Rv32IsEqualModAdapterCols<F, NUM_READS, BLOCKS_PER_READ, BLOCK_SIZE> =
+            adapter_row.borrow_mut();
+        let mut timestamp = cols.from_state.timestamp.as_canonical_u32();
+        let mut timestamp_pp = || {
+            timestamp += 1;
+            timestamp - 1
+        };
+
+        cols.rs_read_aux.iter_mut().for_each(|aux| {
+            mem_helper.fill_from_prev(timestamp_pp(), aux.as_mut());
+        });
+
+        cols.heap_read_aux.iter_mut().for_each(|reads| {
+            reads
+                .iter_mut()
+                .for_each(|aux| mem_helper.fill_from_prev(timestamp_pp(), aux.as_mut()));
+        });
+
+        mem_helper.fill_from_prev(timestamp_pp(), cols.writes_aux.as_mut());
+
+        // Range checks:
+        let need_range_check: Vec<u32> = cols
```

Comment:
for performance, try to avoid allocating new memory: ok if you have to hardcode some indices, but it should be possible to just do the for loop on `request_range` without allocation